database/sqlite3.ksy

```diff
@@ -27,218 +27,487 @@ doc: |
-    type: u1
-    enum: versions
-  - id: read_version
-    type: u1
-    enum: versions
-  - id: reserved_space
-    type: u1
-    doc: Bytes of unused "reserved" space at the end of each page. Usually 0.
-  - id: max_payload_frac
-    type: u1
-    doc: Maximum embedded payload fraction. Must be 64.
-  - id: min_payload_frac
-    type: u1
-    doc: Minimum embedded payload fraction. Must be 32.
-  - id: leaf_payload_frac
-    type: u1
-    doc: Leaf payload fraction. Must be 32.
-  - id: file_change_counter
-    type: u4
-  - id: num_pages
-    type: u4
-    doc: Size of the database file in pages. The "in-header database size".
-  - id: first_freelist_trunk_page
-    type: u4
-    doc: Page number of the first freelist trunk page.
-  - id: num_freelist_pages
-    type: u4
-    doc: Total number of freelist pages.
-  - id: schema_cookie
-    type: u4
-  - id: schema_format
-    type: u4
-    doc: The schema format number. Supported schema formats are 1, 2, 3, and 4.
-  - id: def_page_cache_size
-    type: u4
-    doc: Default page cache size.
-  - id: largest_root_page
-    type: u4
-    doc: The page number of the largest root b-tree page when in auto-vacuum or incremental-vacuum modes, or zero otherwise.
-  - id: text_encoding
-    type: u4
-    enum: encodings
-    doc: The database text encoding. A value of 1 means UTF-8. A value of 2 means UTF-16le. A value of 3 means UTF-16be.
-  - id: user_version
-    type: u4
-    doc: The "user version" as read and set by the user_version pragma.
-  - id: is_incremental_vacuum
-    type: u4
-    doc: True (non-zero) for incremental-vacuum mode. False (zero) otherwise.
-  - id: application_id
-    type: u4
-    doc: The "Application ID" set by PRAGMA application_id.
-  - id: reserved
-    size: 20
-  - id: version_valid_for
-    type: u4
-  - id: sqlite_version_number
-    type: u4
-  - id: root_page
-    type: btree_page
+  - id: header
+    type: database_header
 instances:
-  len_page:
-    value: 'len_page_mod == 1 ? 0x10000 : len_page_mod'
+  pages:
+    type: page(_index + 1, header.page_size * _index)
+    repeat: expr
+    repeat-expr: header.num_pages
 types:
+  page:
+    params:
+      - id: page_number
+        type: s4
+      - id: ofs_body
+        type: s4
+    instances:
+      page_index:
+        value: 'page_number - 1'
+      body:
+        pos: ofs_body
+        size: _root.header.page_size
+        type:
+          switch-on: '(page_index == _root.header.idx_lock_byte_page ? 0 : page_index >= _root.header.idx_first_ptrmap_page and page_index <= _root.header.idx_last_ptrmap_page ? 1 : 2)'
+          cases:
+            0: lock_byte_page(page_number)
+            1: ptrmap_page(page_number)
+            # TODO: Free pages and cell overflow pages are incorrectly interpreted as btree pages
+            # This is unfortunate, but unavoidable since there's no way to recognize these types at
+            # this point in the parser.
+            2: btree_page(page_number)
```

Comment:
@generalmimon no luck with lazy `db.pages`

this still loops all pages when i read `db.pages[0]`

```py
    @property
    def pages(self):
        if hasattr(self, "_m_pages"):
            return self._m_pages

        self._m_pages = []
        for i in range(self.header.num_pages):
            self._m_pages.append(
                Sqlite3.Page(
                    (i + 1), (self.header.page_size * i), self._io, self, self._root
                )
            )
```

Reply:
@milahu:

> this still loops all pages when i read `db.pages[0]`

Yes, it creates the objects but does not parse them.

If you don't want to even create the empty objects (because the total memory usage of too many empty objects would be too high), you can provide an "unused" `page` type and let the users of the parser instantiate it themselves for the page number they want (and even dispose the object afterwards to keep the memory usage low). This approach is essentially explained in https://stackoverflow.com/a/73332294/12940655.

KS unfortunately doesn't support truly unused types very well at the moment (i.e. when you define a type in the `types` but don't use it absolutely anywhere), but this can be easily worked around by the `if: false` trick as explained in the linked SO post.